File: bgm-code/ParameterEstimation_Stat/ExamplesBinomial_Estimation/Rate_4.R
# clears workspace:  
rm(list=ls(all=TRUE)) 

# sets working directories:
#setwd("F:/Codes/ExamplesBinomial_Estimation")
setwd("C:/Documents and Settings/ewagenmakers/My Documents/My Dropbox/EJ/temp/BayesBook/Code/ParameterEstimation_Stat/ExamplesBinomial_Estimation")

library(R2WinBUGS)
bugsdir = "C:/Program Files/WinBUGS14"

k = c(3, 1, 4, 5)
n = 10
m = length(k)

data  = list("k", "n", "m") # to be passed on to WinBUGS
myinits =	list(
  list(theta = 0.5, thetaprior = 0.5))

# parameters to be monitored:	
parameters = c("theta", "thetaprior", "postpredk", "priorpredk")

# The following command calls WinBUGS with specific options.
# For a detailed description see Sturtz, Ligges, & Gelman (2005).
samples = bugs(data, inits=myinits, parameters,
	 			model.file ="Rate_4.txt",
	 			n.chains=1, n.iter=5000, n.burnin=0, n.thin=1,
	 			DIC=T, bugs.directory=bugsdir,
	 			codaPkg=F, debug=T)
# Now the values for the monitored parameters are in the "samples" object, 
# ready for inspection.

######################Plots########################################################################
theta      = samples$sims.list$theta
thetaprior = samples$sims.list$thetaprior 
postpredk  = samples$sims.list$postpredk 
priorpredk = samples$sims.list$priorpredk

layout(matrix(c(1,2),2,1))
layout.show(2)
#Prior and posterior of theta
plot(density(theta), zero.line=F, axes=F, main="", xlab="", ylab="", xlim=c(0,1), ylim=c(0,6))
axis(1, at=c(0,0.2,0.4,0.6,0.8,1), lab=c("0","0.2","0.4","0.6","0.8","1"),cex.axis=0.8)
mtext("Rate", side=1, line=2.25, cex=1.2)
axis(2, at=c(0,2,4,6),cex.axis=0.8)
mtext("Posterior Density", side=2, line=2.25, cex=1.2)
lines(density(thetaprior, from=0, to=1), lty=3, col="gray")
legend(0.775,5.75, c("Prior", "Posterior"), lty=c(3,1), col=c ("grey", "black"))

#Prior and posterior predictive
hist(postpredk,breaks=c(-0.5,0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5,9.5,10.5),freq=F, right=F, ylab="", xlab="", ylim=c(0,0.3),main="", axes=F )
axis(1, at=c(0,2,4,6,8,10),lab=c(0,2,4,6,8,10),cex.axis=0.8)
mtext("Success Count", side=1, line=2.25, cex=1.2)
axis(2,at=c(0,0.1,0.2,0.3),lab=c("0","0.1","0.2","0.3"),cex.axis=0.8)
mtext("Posterior Mass", side=2, line=2.25, cex=1.2)
hist(priorpredk, breaks=c(-0.5,0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5,9.5,10.5),freq=F,right=F,add=T, lty=3,border="grey")
legend(8,0.3, c("Prior", "Posterior"), lty=c(3,1),col=c("grey", "black"))

File: bgm-code/ParameterEstimation_Stat/ExamplesBinomial_Estimation/Rate_4.m
% Prior and Posterior Predictive

clear;

% Data 
k=[3 1 4 5];n=10;
% Derived Variables
m=length(k);

% WinBUGS Parameters
nchains=1; % How Many Chains?
nburnin=0; % How Many Burn-in Samples?
nsamples=5e3; % How Many Recorded Samples?

% Assign Matlab Variables to the Observed WinBUGS Nodes
datastruct = struct('k',k,'n',n,'m',m);

% Initialize Unobserved Variables
for i=1:nchains
    S.theta = 0.5; % Intial Value
    init0(i) = S;
end

% Use WinBUGS to Sample
[samples, stats] = matbugs(datastruct, ...
    fullfile(pwd, 'Rate_4.txt'), ...
    'init', init0, ...
    'nChains', nchains, ...
    'view', 1, 'nburnin', nburnin, 'nsamples', nsamples, ...
    'thin', 1, 'DICstatus', 0, 'refreshrate',100, ...
    'monitorParams', {'theta','thetaprior','postpredk','priorpredk'}, ...
    'Bugdir', 'C:/Program Files/WinBUGS14');

% Analysis
figure(1);clf;
% Parameter Space
subplot(211);hold on;
eps=.01;bins=[eps:eps:1-eps];
count=hist(samples.thetaprior,bins);
count=count/sum(count)/eps;
ph=plot(bins,count,'k--');
count=hist(samples.theta,bins);
count=count/sum(count)/eps;
ph=plot(bins,count,'k-');
legend('Prior','Posterior');
set(gca,'box','on','fontsize',14);
xlabel('Rate','fontsize',16);
ylabel('Posterior Density','fontsize',16);
% Data Space
subplot(212);hold on;
kbins=[0:n];
count1=histc(samples.priorpredk,kbins);
count1=count1/sum(count1);
ph=bar(kbins,count1,.6);
set(ph,'facecolor','none','linewidth',1.5,'linestyle','--')
count2=histc(samples.postpredk,kbins);
count2=count2/sum(count2);
ph=bar(kbins,count2);set(ph,'facecolor','none')
%ph=bar(kbins,[count1;count2]');
legend('Prior','Posterior');
set(gca,'xlim',[-1 n+1],'box','on','fontsize',14);
xlabel('Success Count','fontsize',16);
ylabel('Posterior Mass','fontsize',16);